frame/election-provider-multi-phase/test-election-provider/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+mod mock;
+
+pub(crate) const LOG_TARGET: &str = "tests::epm";
+
+use mock::*;
+use sp_core::Get;
+use sp_npos_elections::{to_supports, StakedAssignment};
+
+use crate::mock::RuntimeOrigin;
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("🛠️  ", $patter)  $(, $values)*
+        )
+    };
+}
+
+fn log_current_time() {
+	log!(
+		trace,
+		"block: {:?}, session: {:?}, era: {:?}, EPM phase: {:?} ts: {:?}",
+		System::block_number(),
+		Session::current_index(),
+		Staking::current_era(),
+		ElectionProviderMultiPhase::current_phase(),
+		Timestamp::now()
+	);
+}
+
+#[test]
+/// Replicates the Kusama incident of 8th Dec 2022 and its resolution through the governance
+/// fallback.
+///
+/// After enough slashes to exceed the `Staking::OffendingValidatorsThreshold`, the staking pallet
+/// set `Forcing::ForceNew`. When a new session starts, staking will start to force a new era and
+/// calls <EPM as election_provider>::elect(). If at this point EPM and the staking miners did not
+/// have enough time to queue a new solution (snapshot + solution submission), the election request
+/// fails. If there is no election fallback mechanism in place, EPM enters in emergency mode.
+/// Recovery: Once EPM is in emergency mode, subsequent calls to `elect()` will fail until a new
+/// solution is added to EPM's `QueuedSolution` queue. This can be achieved through
+/// `Call::set_emergency_election_result` or `Call::governance_fallback` dispatchables. Once a new
+/// solution is added to the queue, EPM phase transitions to `Phase::Off` and the election flow
+/// restarts.
+fn enters_emergency_phase_after_forcing_before_elect() {
+	ExtBuilder::default().initialize_first_session(true).build_and_execute(|| {
+		log!(
+			trace,
+			"current validators (staking): {:?}",
+			<Runtime as pallet_staking::SessionInterface<AccountId>>::validators()
+		);
+		let session_validators_before = Session::validators();
+
+		roll_to_epm_off();
+		assert!(ElectionProviderMultiPhase::current_phase().is_off());
+
+		assert_eq!(pallet_staking::ForceEra::<Runtime>::get(), pallet_staking::Forcing::NotForcing);
+		// slashes until staking gets into `Forcing::ForceNew`.
+		add_slash(&11);
+
+		assert_eq!(pallet_staking::ForceEra::<Runtime>::get(), pallet_staking::Forcing::ForceNew);
+
+		advance_session();
```

Comment:
Also, we should plan the slashes to happen EXACTLY at the block where if we do `roll_to(current_block + 1)`, a new session would happen automatically. 

Right now, you are kinda simulating that 🙈 

Not sure how accurate you want to be. Perhaps I am too strict here. 

But, the nice thing is:

Assume the block at which the slashes happen is `n`, and the block at which the next session+election happens in `x`. If the difference between `n` and `x` is too short, then we fail. But if we have like 5 block between the two, this would not happen. 

As a side note, in order to make these tests more accurate, you can inspire from here to also have automatic submission of OCWs: https://github.com/paritytech/substrate/blob/kiz-multi-block-election/frame/election-provider-multi-block/src/mock/mod.rs#L580

This could help tests be more accurate.

Reply:
Ah cool! Yes, I just finished my own flavour of `fn roll_with_ocw`, I should have looked a bit more into the existing  code. But yes, I agree. The next commit should have a more accurate flow without simulating emergency etc